# Module 10 Application

## Challenge: Crypto Clustering

In this Challenge, you’ll combine your financial Python programming skills with the new unsupervised learning skills that you acquired in this module.

The CSV file provided for this challenge contains price change data of cryptocurrencies in different periods.

The steps for this challenge are broken out into the following sections:

* Import the Data (provided in the starter code)
* Prepare the Data (provided in the starter code)
* Find the Best Value for `k` Using the Original Data
* Cluster Cryptocurrencies with K-means Using the Original Data
* Optimize Clusters with Principal Component Analysis
* Find the Best Value for `k` Using the PCA Data
* Cluster the Cryptocurrencies with K-means Using the PCA Data
* Visualize and Compare the Results

### Import the Data

This section imports the data into a new DataFrame. It follows these steps:

1. Read  the “crypto_market_data.csv” file from the Resources folder into a DataFrame, and use `index_col="coin_id"` to set the cryptocurrency name as the index. Review the DataFrame.

2. Generate the summary statistics, and use HvPlot to visualize your data to observe what your DataFrame contains.


> **Rewind:** The [Pandas`describe()`function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) generates summary statistics for a DataFrame. 

In [78]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [79]:
# import bread and butter
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', 1000)
pd.set_option('display.width', 1000)
from watermark import watermark
print(watermark())
print(watermark(iversions=True, globals_=globals(), packages=""))

ModuleNotFoundError: No module named 'watermark'

In [80]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    Path("Resources/crypto_market_data.csv"),
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  price_change_percentage_14d  price_change_percentage_30d  price_change_percentage_60d  price_change_percentage_200d  price_change_percentage_1y
coin_id                                                                                                                                                                                                                  
bitcoin                              1.08388                     7.60278                      6.57509                      7.67258                     -3.25185                      83.51840                    37.51761
ethereum                             0.22392                    10.38134                      4.80849                      0.13169                    -12.88890                     186.77418                   101.96023
tether                              -0.21173                     0.04935                      0.00640                     -0.04237                      0.28037                      -0.00542                     0.01954
ripple                              -0.37819                    -0.60926                      2.24984                      0.23455                    -17.55245                      39.53888                   -16.60193
bitcoin-cash                         2.90585                    17.09717                     14.75334                     15.74903                    -13.71793                      21.66042                    14.49384
binancecoin                          2.10423                    12.85511                      6.80688                      0.05865                     36.33486                     155.61937                    69.69195
chainlink                           -0.23935                    20.69459                      9.30098                    -11.21747                    -43.69522                     403.22917                   325.13186
cardano                              0.00322                    13.99302                      5.55476                     10.10553                    -22.84776                     264.51418                   156.09756
litecoin                            -0.06341                     6.60221                      7.28931                      1.21662                    -17.23960                      27.49919                   -12.66408
bitcoin-cash-sv                      0.92530                     3.29641                     -1.86656                      2.88926                    -24.87434                       7.42562                    93.73082

In [81]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  price_change_percentage_14d  price_change_percentage_30d  price_change_percentage_60d  price_change_percentage_200d  price_change_percentage_1y
count                    41.000000                   41.000000                    41.000000                    41.000000                    41.000000                     41.000000                   41.000000
mean                     -0.269686                    4.497147                     0.185787                     1.545693                    -0.094119                    236.537432                  347.667956
std                       2.694793                    6.375218                     8.376939                    26.344218                    47.365803                    435.225304                 1247.842884
min                     -13.527860                   -6.094560                   -18.158900                   -34.705480                   -44.822480                     -0.392100                  -17.567530
25%                      -0.608970                    0.047260                    -5.026620                   -10.438470                   -25.907990                     21.660420                    0.406170
50%                      -0.063410                    3.296410                     0.109740                    -0.042370                    -7.544550                     83.905200                   69.691950
75%                       0.612090                    7.602780                     5.510740                     4.578130                     0.657260                    216.177610                  168.372510
max                       4.840330                   20.694590                    24.239190                   140.795700                   223.064370                   2227.927820                 7852.089700

In [82]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

This section prepares the data before running the K-Means algorithm. It follows these steps:

1. Use the `StandardScaler` module from scikit-learn to normalize the CSV file data. This will require you to utilize the `fit_transform` function.

2. Create a DataFrame that contains the scaled data. Be sure to set the `coin_id` index from the original DataFrame as the index for the new DataFrame. Review the resulting DataFrame.


In [83]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data)

In [84]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns
)

# Copy the crypto names from the original data
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  price_change_percentage_14d  price_change_percentage_30d  price_change_percentage_60d  price_change_percentage_200d  price_change_percentage_1y
coin_id                                                                                                                                                                                                               
bitcoin                          0.508529                    0.493193                     0.772200                     0.235460                    -0.067495                     -0.355953                   -0.251637
ethereum                         0.185446                    0.934445                     0.558692                    -0.054341                    -0.273483                     -0.115759                   -0.199352
tether                           0.021774                   -0.706337                    -0.021680                    -0.061030                     0.008005                     -0.550247                   -0.282061
ripple                          -0.040764                   -0.810928                     0.249458                    -0.050388                    -0.373164                     -0.458259                   -0.295546
bitcoin-cash                     1.193036                    2.000959                     1.760610                     0.545842                    -0.291203                     -0.499848                   -0.270317

---

### Find the Best Value for k Using the Original Data

In this section, you will use the elbow method to find the best value for `k`.

1. Code the elbow method algorithm to find the best value for `k`. Use a range from 1 to 11. 

2. Plot a line chart with all the inertia values computed with the different values of `k` to visually identify the optimal value for `k`.

3. Answer the following question: What is the best value for `k`?

In [85]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
# YOUR CODE HERE!
k = list(range(1,11))

In [86]:
# Create an empy list to store the inertia values
# YOUR CODE HERE!
inertia = []


In [87]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
# YOUR CODE HERE!
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)

In [88]:
# Create a dictionary with the data to plot the Elbow curve
# YOUR CODE HERE!
elbow_dict = {'k': k, 'inertia': inertia}
# Create a DataFrame with the data to plot the Elbow curve
# YOUR CODE HERE!
df_elbow = pd.DataFrame(elbow_dict)

In [89]:
df_elbow

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.302379
5   6   52.888518
6   7   43.914690
7   8   37.517032
8   9   32.485241
9  10   28.222899

In [90]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# YOUR CODE HERE!
df_elbow.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve',
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: What is the best value for k?
**Question:** What is the best value for `k`?

**Answer:** # YOUR ANSWER HERE!

7

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In this section, you will use the K-Means algorithm with the best value for `k` found in the previous section to cluster the cryptocurrencies according to the price changes of cryptocurrencies provided.

1. Initialize the K-Means model with four clusters using the best value for `k`. 

2. Fit the K-Means model using the original data.

3. Predict the clusters to group the cryptocurrencies using the original data. View the resulting array of cluster values.

4. Create a copy of the original data and add a new column with the predicted clusters.

5. Create a scatter plot using hvPlot by setting `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.

In [91]:
# Initialize the K-Means model using the best value for k
# YOUR CODE HERE!
model = KMeans(n_clusters=7, random_state=1)


In [92]:
# Fit the K-Means model using the scaled data
# YOUR CODE HERE!
model.fit(df_market_data_scaled)

KMeans(n_clusters=7, random_state=1)

In [93]:
# Predict the clusters to group the cryptocurrencies using the scaled data
# YOUR CODE HERE!
crypto_groupings = model.predict(df_market_data_scaled)

# View the resulting array of cluster values.
# YOUR CODE HERE!
crypto_groupings

array([1, 1, 4, 4, 6, 1, 6, 1, 1, 4, 0, 4, 4, 6, 4, 1, 0, 4, 1, 4, 0, 1,
       4, 4, 4, 4, 4, 0, 1, 5, 4, 4, 3, 1, 0, 4, 2, 0, 4, 4, 4],
      dtype=int32)

In [94]:
# Create a copy of the DataFrame
# YOUR CODE HERE!
df_market_data_scaled_copy = df_market_data_scaled.copy()

In [95]:
# Add a new column to the DataFrame with the predicted clusters
# YOUR CODE HERE!
df_market_data_scaled_copy['predicted clusters'] = crypto_groupings

# Display sample data
# YOUR CODE HERE!
df_market_data_scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  price_change_percentage_14d  price_change_percentage_30d  price_change_percentage_60d  price_change_percentage_200d  price_change_percentage_1y  predicted clusters
coin_id                                                                                                                                                                                                                                   
bitcoin                          0.508529                    0.493193                     0.772200                     0.235460                    -0.067495                     -0.355953                   -0.251637                   1
ethereum                         0.185446                    0.934445                     0.558692                    -0.054341                    -0.273483                     -0.115759                   -0.199352                   1
tether                           0.021774                   -0.706337                    -0.021680                    -0.061030                     0.008005                     -0.550247                   -0.282061                   4
ripple                          -0.040764                   -0.810928                     0.249458                    -0.050388                    -0.373164                     -0.458259                   -0.295546                   4
bitcoin-cash                     1.193036                    2.000959                     1.760610                     0.545842                    -0.291203                     -0.499848                   -0.270317                   6

In [96]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# YOUR CODE HERE!
df_market_data_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    hover_cols="index",
    by="predicted clusters"
)

:NdOverlay   [predicted clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis

In this section, you will perform a principal component analysis (PCA) and reduce the features to three principal components.

1. Create a PCA model instance and set `n_components=3`.

2. Use the PCA model to reduce to three principal components. View the first five rows of the DataFrame. 

3. Retrieve the explained variance to determine how much information can be attributed to each principal component.

4. Answer the following question: What is the total explained variance of the three principal components?

5. Create a new DataFrame with the PCA data. Be sure to set the `coin_id` index from the original DataFrame as the index for the new DataFrame. Review the resulting DataFrame.

In [97]:
# Create a PCA model instance and set `n_components=3`.
# YOUR CODE HERE!
pca=PCA(n_components=3)

In [110]:
df_market_data_scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  price_change_percentage_14d  price_change_percentage_30d  price_change_percentage_60d  price_change_percentage_200d  price_change_percentage_1y  predicted clusters
coin_id                                                                                                                                                                                                                                   
bitcoin                          0.508529                    0.493193                     0.772200                     0.235460                    -0.067495                     -0.355953                   -0.251637                   1
ethereum                         0.185446                    0.934445                     0.558692                    -0.054341                    -0.273483                     -0.115759                   -0.199352                   1
tether                           0.021774                   -0.706337                    -0.021680                    -0.061030                     0.008005                     -0.550247                   -0.282061                   4
ripple                          -0.040764                   -0.810928                     0.249458                    -0.050388                    -0.373164                     -0.458259                   -0.295546                   4
bitcoin-cash                     1.193036                    2.000959                     1.760610                     0.545842                    -0.291203                     -0.499848                   -0.270317                   6

In [98]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
# YOUR CODE HERE!
data_pca = pca.fit_transform(df_market_data_scaled_copy)

# View the first five rows of the DataFrame. 
# YOUR CODE HERE!
data_pca[:5]

array([[-1.26076912, -0.60681217,  1.44623335],
       [-1.4063869 , -0.45739797,  1.08240841],
       [ 1.03217742, -0.43641073, -0.6352075 ],
       [ 1.02555692, -0.4743131 , -0.70151687],
       [ 3.82755791, -1.21404715,  0.45784917]])

In [99]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
# YOUR CODE HERE!
pca.explained_variance_ratio_

array([0.35919943, 0.24825632, 0.20658928])

In [100]:
pca.explained_variance_ratio_.sum()

0.8140450235455852

#### Answer the following question: What is the total explained variance of the three principal components?

**Question:** What is the total explained variance of the three principal components?

**Answer:** # YOUR ANSWER HERE!

0.8140450235455852


In [122]:
columns = df_market_data_scaled_copy.index.tolist()
columns

['bitcoin',
 'ethereum',
 'tether',
 'ripple',
 'bitcoin-cash',
 'binancecoin',
 'chainlink',
 'cardano',
 'litecoin',
 'bitcoin-cash-sv',
 'crypto-com-chain',
 'usd-coin',
 'eos',
 'monero',
 'tron',
 'tezos',
 'okb',
 'stellar',
 'cosmos',
 'cdai',
 'neo',
 'wrapped-bitcoin',
 'leo-token',
 'huobi-token',
 'nem',
 'binance-usd',
 'iota',
 'vechain',
 'zcash',
 'theta-token',
 'dash',
 'ethereum-classic',
 'ethlend',
 'maker',
 'havven',
 'omisego',
 'celsius-degree-token',
 'ontology',
 'ftx-token',
 'true-usd',
 'digibyte']

In [131]:
# Create a new DataFrame with the PCA data.
# Note: The code for this step is provided for you
columns = df_market_data_scaled_copy.index.tolist()

# Creating a DataFrame with the PCA data
# YOUR CODE HERE!
data_pca = pd.DataFrame(
    data_pca,
    columns=["PCA1", "PCA2", "PCA3"],
    # index_col=df_market_data_scaled_copy['coin_id']
)
# Copy the crypto names from the original data
# YOUR CODE HERE!
# columns = df_market_data_scaled_copy.index.tolist()

# Set the coinid column as index
# YOUR CODE HERE!
data_pca['coinid'] = columns
# Display sample data
# YOUR CODE HERE!
data_pca = data_pca.set_index('coinid')
data_pca

PCA1      PCA2      PCA3
coinid                                            
bitcoin              -1.260769 -0.606812  1.446233
ethereum             -1.406387 -0.457398  1.082408
tether                1.032177 -0.436411 -0.635208
ripple                1.025557 -0.474313 -0.701517
bitcoin-cash          3.827558 -1.214047  0.457849
binancecoin          -1.044933 -0.532819  1.946336
chainlink             3.239575 -0.479007 -0.956599
cardano              -1.286621 -0.349927  1.315853
litecoin             -1.406521 -0.648167  1.050396
bitcoin-cash-sv       1.009711 -0.761669 -0.672819
crypto-com-chain     -3.156089 -0.206667 -0.069178
usd-coin              1.028637 -0.441606 -0.641061
eos                   0.917089 -0.691146 -0.930071
monero                4.195698 -0.012121  1.273196
tron                  1.073036 -0.398014 -0.602791
tezos                -1.780359 -0.777673  0.191169
okb                  -3.063798  0.103118 -0.011488
stellar               0.819830 -0.482043 -1.175477
cosmos               -1.284145 -0.310257  1.278458
cdai                  1.040053 -0.517310 -0.611726
neo                  -2.995176 -0.327919  0.269435
wrapped-bitcoin      -1.267443 -0.610117  1.432845
leo-token             0.785753 -0.406808 -1.056005
huobi-token           1.023903 -0.410113 -0.687762
nem                   1.329189  0.591979  0.067603
binance-usd           1.038324 -0.453861 -0.619245
iota                  0.883713 -0.761484 -0.971078
vechain              -3.433461 -0.503869 -0.552753
zcash                -1.379570 -0.425057  1.102073
theta-token           1.962551  2.665466 -0.762107
dash                  0.937789 -0.611831 -0.966175
ethereum-classic      0.962236 -0.579856 -0.824793
ethlend              -1.166964  8.157636 -3.656278
maker                -1.562354 -0.382345  0.855135
havven               -3.466546  0.923098 -0.919256
omisego               1.143352  0.098435  0.042092
celsius-degree-token  2.002328  4.670759  6.672456
ontology             -3.424467 -0.577500 -0.784096
ftx-token             1.117576 -0.600148 -0.486114
true-usd              1.044538 -0.462066 -0.605678
digibyte              0.945429 -0.300106 -0.584264

---

### Find the Best Value for k Using the PCA Data

In this section, you will use the elbow method to find the best value for `k` using the PCA data.

1. Code the elbow method algorithm and use the PCA data to find the best value for `k`. Use a range from 1 to 11. 

2. Plot a line chart with all the inertia values computed with the different values of `k` to visually identify the optimal value for `k`.

3. Answer the following questions: What is the best value for k when using the PCA data? Does it differ from the best k value found using the original data?

In [133]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
# YOUR CODE HERE!
k = list(range(1,11))

In [134]:
# Create an empy list to store the inertia values
# YOUR CODE HERE!
inertia = []

In [ ]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
# YOUR CODE HERE!


In [ ]:
# Create a dictionary with the data to plot the Elbow curve
# YOUR CODE HERE!

# Create a DataFrame with the data to plot the Elbow curve
# YOUR CODE HERE!

In [ ]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# YOUR CODE HERE!

#### Answer the following questions: What is the best value for k when using the PCA data? Does it differ from the best k value found using the original data?
* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** # YOUR ANSWER HERE!


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** # YOUR ANSWER HERE!

---

### Cluster Cryptocurrencies with K-means Using the PCA Data

In this section, you will use the PCA data and the K-Means algorithm with the best value for `k` found in the previous section to cluster the cryptocurrencies according to the principal components.

1. Initialize the K-Means model with four clusters using the best value for `k`. 

2. Fit the K-Means model using the PCA data.

3. Predict the clusters to group the cryptocurrencies using the PCA data. View the resulting array of cluster values.

4. Add a new column to the DataFrame with the PCA data to store the predicted clusters.

5. Create a scatter plot using hvPlot by setting `x="PC1"` and `y="PC2"`. Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.

In [ ]:
# Initialize the K-Means model using the best value for k
# YOUR CODE HERE!

In [ ]:
# Fit the K-Means model using the PCA data
# YOUR CODE HERE!

In [ ]:
# Predict the clusters to group the cryptocurrencies using the PCA data
# YOUR CODE HERE!

# View the resulting array of cluster values.
# YOUR CODE HERE!

In [ ]:
# Create a copy of the DataFrame with the PCA data
# YOUR CODE HERE!

# Add a new column to the DataFrame with the predicted clusters
# YOUR CODE HERE!

# Display sample data
# YOUR CODE HERE!

In [ ]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# YOUR CODE HERE!

---

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

1. Create a composite plot using hvPlot and the plus (`+`) operator to contrast the Elbow Curve that you created to find the best value for `k` with the original and the PCA data.

2. Create a composite plot using hvPlot and the plus (`+`) operator to contrast the cryptocurrencies clusters using the original and the PCA data.

3. Answer the following question: After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

> **Rewind:** Back in Lesson 3 of Module 6, you learned how to create composite plots. You can look at that lesson to review how to make these plots; also, you can check [the hvPlot documentation](https://holoviz.org/tutorial/Composing_Plots.html).

In [ ]:
# Composite plot to contrast the Elbow curves
# YOUR CODE HERE!

In [ ]:
# Compoosite plot to contrast the clusters
# YOUR CODE HERE!

#### Answer the following question: After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** # YOUR ANSWER HERE!